### DISCUSSION TITLE
How to set a default for the entire struct, instead of for each field?

### DISCUSSION THREAD

**Discussion by rain2307** [elastio/bon#151] [Q&A] [answered]
current: 

```rust
#[derive(Debug, Clone, Builder, Default)]
#[builder]
pub struct A {
  #[builder(default)] // here
  pub name: String,
}
```

How to do this?

```rust
#[derive(Debug, Clone, Builder, Default)]
#[builder(default)] // struct level default
pub struct A {
  pub name: String,
}
```

**Comment by Veetaha** [comment, marked as answer]
Hi! Thank you for opening a discussion.  The short answer is that currently there is no syntax to set `default` at the top level. However, it's something that I'm planning to add in the next several releases. I've been thinking on the design of this feature trying to select the most intuitive behavior for this. 

I've created an issue with the anticipated design for this. I'd be glad to hear any feedback about it: https://github.com/elastio/bon/issues/152

**Reply by rain2307** [reply]
Nice! 
I have some ideas maybe useful 

1. It is enabled by default if there is #[Default]
2. If the field type is u8 or ...,default is automatically added.
2. We can set a function to decide the default value

Thanks for your work!

**Reply by Veetaha** [reply]
Regarding the first two points: unfortunately the macro has no information of wether the type implements `Default` or not. The macro only sees the raw text it was given and it can't query the compiler wether the given type implements some trait. Making a member default automatically doesn't always make sense, therefore it requires explicit annotation today.

As for the 3-rd point - it is already possible to do today. You can specify a custom default value like this: 
`#[builder(default = 2 + 2)]` where you could also call any function: `#[builder(default = any_function())]`
